tpm/key.go

```diff
@@ -0,0 +1,505 @@
+	return k.attestedBy == ak.name
+}
+
+// Certificate returns the certificate for the Key, if set.
+// Will return nil in case no AK certificate is available.
+func (k *Key) Certificate() *x509.Certificate {
+	if len(k.chain) == 0 {
+		return nil
+	}
+	return k.chain[0]
+}
+
+// CertificateChain returns the certificate chain for the Key.
+// It can return an empty chain.
+func (k *Key) CertificateChain() []*x509.Certificate {
+	return k.chain
+}
+
+// CreatedAt returns the the creation time of the Key.
+func (k *Key) CreatedAt() time.Time {
+	return k.createdAt.Truncate(time.Second)
+}
+
+// MarshalJSON marshals the Key to JSON.
+func (k *Key) MarshalJSON() ([]byte, error) {
+	chain := make([][]byte, len(k.chain))
+	for i, cert := range k.chain {
+		chain[i] = cert.Raw
+	}
+	o := struct {
+		Name       string    `json:"name"`
+		Data       []byte    `json:"data"`
+		AttestedBy string    `json:"attestedBy,omitempty"`
+		Chain      [][]byte  `json:"chain,omitempty"`
+		CreatedAt  time.Time `json:"createdAt"`
+	}{
+		Name:       k.name,
+		Data:       k.data,
+		AttestedBy: k.attestedBy,
+		Chain:      chain,
+		CreatedAt:  k.createdAt,
+	}
+	return json.Marshal(o)
+}
+
+// comparablePublicKey is an interface that allows an crypto.PublicKey to be
+// compared to another crypto.PublicKey.
+type comparablePublicKey interface {
+	Equal(crypto.PublicKey) bool
+}
+
+// CreateKeyConfig is used to pass configuration
+// when creating Keys.
+type CreateKeyConfig struct {
+	// Algorithm to be used, either RSA or ECDSA.
+	Algorithm string
+	// Size is used to specify the bit size of the key or elliptic curve. For
+	// example, '256' is used to specify curve P-256.
+	Size int
+
+	// TODO(hs): move key name to this struct?
+}
+
+// AttestKeyConfig is used to pass configuration
+// when creating Keys that are to be attested by
+// an AK.
+type AttestKeyConfig struct {
+	// Algorithm to be used, either RSA or ECDSA.
+	Algorithm string
+	// Size is used to specify the bit size of the key or elliptic curve. For
+	// example, '256' is used to specify curve P-256.
+	Size int
+	// QualifyingData is additional data that is passed to the TPM.
+	// It can be used as a nonce to ensure freshness of an attestation.
+	// When used with ACME `device-attest-01`, this contains a hash of
+	// the key authorization.
+	QualifyingData []byte
+
+	// TODO(hs): add akName and key name to this struct?
+}
+
+// CreateKey creates a new Key identified by `name`. If no name is  provided,
+// a random 10 character name is generated. If a Key with the same name exists,
+// `ErrExists` is returned. The Key won't be attested by an AK.
+func (t *TPM) CreateKey(ctx context.Context, name string, config CreateKeyConfig) (*Key, error) {
+	var err error
+	if err = t.open(ctx); err != nil {
+		return nil, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer func() {
+		if tempErr := t.close(ctx); tempErr != nil && err != nil {
+			err = tempErr
+		}
+	}()
+
+	now := time.Now().UTC()
```

Comment:
Please do check around the codebase to make sure that all date handling ends up being either all UTC or all Local and not mixed, please.